src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusCollectionManager.cs

```diff
@@ -0,0 +1,216 @@
+                }
+
+                Interlocked.Increment(ref this.readerCount);
+                this.ExitGlobalLock();
+#if NETCOREAPP3_1_OR_GREATER
+                return new ValueTask<ArraySegment<byte>>(this.collectionTcs.Task);
+#else
+                return this.collectionTcs.Task;
+#endif
+            }
+
+            this.WaitForReadersToComplete();
+
+            // Start a collection on the current thread.
+            this.collectionRunning = true;
+            this.previousDataViewExpirationAtUtc = null;
+            Interlocked.Increment(ref this.readerCount);
+            this.ExitGlobalLock();
+
+            bool result = this.ExecuteCollect();
+            if (result && this.scrapeResponseCacheDurationInMilliseconds > 0)
+            {
+                this.previousDataViewExpirationAtUtc = DateTime.UtcNow.AddMilliseconds(this.scrapeResponseCacheDurationInMilliseconds);
+            }
+
+            this.EnterGlobalLock();
+
+            this.collectionRunning = false;
+
+            if (this.collectionTcs != null)
+            {
+                this.collectionTcs.SetResult(this.previousDataView);
+                this.collectionTcs = null;
+            }
+
+            this.ExitGlobalLock();
+
+#if NETCOREAPP3_1_OR_GREATER
+            return new ValueTask<ArraySegment<byte>>(this.previousDataView);
+#else
+            return Task.FromResult(this.previousDataView);
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void ExitCollect()
+        {
+            Interlocked.Decrement(ref this.readerCount);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnterGlobalLock()
+        {
+            SpinWait lockWait = default;
+            while (true)
+            {
+                if (Interlocked.CompareExchange(ref this.globalLockState, 1, this.globalLockState) != 0)
+                {
+                    lockWait.SpinOnce();
+                    continue;
+                }
+
+                break;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ExitGlobalLock()
+        {
+            this.globalLockState = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WaitForReadersToComplete()
+        {
+            SpinWait readWait = default;
+            while (true)
+            {
+                if (Interlocked.CompareExchange(ref this.readerCount, 0, this.readerCount) != 0)
+                {
+                    readWait.SpinOnce();
+                    continue;
+                }
+
+                break;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteCollect()
+        {
+            this.exporter.OnExport = this.onCollectRef;
```

Comment:
`this.exporter.OnExport = this.OnCollect;`
Would this not have worked?

Reply:
So what this is doing is preventing the allocation of a delegate. The only sure-fire way I know how to do that is using a field. If you do it locally, sometimes the compiler will generate a field and initialize it on the first invocation, other times it will just allocate a delegate for each call. Not totally sure how it decides. I can't remember if this particular case fell into the delegate path or if I was just being cautious. I did do some benchmarks over this when I was working on it so I _probably_ saw allocations and used the field, but I can't remember exactly.

Reply:
Thanks for the explanation, I had a hunch it was for perf reasons